### PR TITLE
Add GEM trigger data to EMTF DAQ data format and unpacker

### PR DESCRIPTION
**Pull request by jsturdy**
#### PR description:

This PR brings upcoming changes to the EMTF firmware for including the GEM trigger payload in the EMTF DAQ data.
The EMTF DAQ module firmware has been updated based on the specifications in DN-20-016, and is currently being integrated into the full EMTF firmware, but no data has yet been taken with the new format.

* Implementation for GE1/1 is fully included
* Implementations for GE2/1 and ME0 are yet to be implemented in FW as the protocol for transmission is not yet finalized

With the changes in this PR, there are no expected changes in current output, as the GEM payload is not in any RAW files and is not present in the MC due to the lack of an EMTF packer (future PR with @jiafulow).

Took the opportunity to reduce usage of magic numbers in places where GEM information was added.

#### PR validation:

As there is no EMTF packer, there is no way to currently emulate this new data stream until either data is taken with the new firmware, or the packer is developed.

Standard PR matrix tests were run on both `CMSSW_11_2_0` and `CMSSW_11_1_0`.
```
runTheMatrix.py -l limited -i all --ibeos
```
reported
```
36 34 30 23 16 4 1 1 1 tests passed, 1 1 3 0 0 0 0 0 0 failed
```
but the failed tests also failed in an unmodified release

### CODE REVIEW

**Comment by jsturdy** [comment]
For information: @eyigitba and @abrinke1

**Comment by silviodonato** [comment]
please test
@rekovic

**Comment by silviodonato** [comment]
kind remind @rekovic

**Comment by boudoul** [comment]
Dear @jsturdy , is this something that would be needed for the next MWGR ? (scheduled first week of September, where the tentative plan is to include GEM) - Thanks

**Comment by jsturdy** [comment]
Hi @boudoul the utility of this is determined by whether or not the EMTF firmware has integrated the GEM input stream into its DAQ module at that time (@jiafulow can probably say whether this has happened or is going to happen prior to the next MWGR, as I have now moved to ATLAS).

In the three cases below, I don't see a problem either including or not including this PR
* If GEMs don't end up participating, the content of this PR is moot
* If GEMs participate in the MWGR but the EMTF DAQ module doesn't include them, again, this PR is moot
* If the EMTF DAQ module includes the GEM data, the old unpacker should still succeed in unpacking the rest of the stream, so if this PR is not included, things *should*[1] still be fine, and once it's included the RAW data can be unpacked with the new unpacker to analyze the GEM hits at the EMTF level.

On the other hand, as soon as this PR is merged and there is data to unpack (whether real or from a simulated/emulated packer), it will probably be useful, and unless there is a regression, it should be fine to get it into a release even before this module is part of the operational EMTF firmware.

[1] Though this last point is not possible to test until there is an input stream including (real or emulated) GEM trigger data in the EMTF DAQ output.

**Comment by jiafulow** [comment]
Yes, we would like to see the PR be merged, but it doesn't necessarily need to happen before the next MWGR. As Jared explained, the existing EMTF unpacker should still work. If it's not merged, we could still run this PR privately to process the RAW data from GEM.

File: DataFormats/L1TMuon/interface/EMTF/GEM.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            gem_bxn(-99),
+            bc0(-99),
+            tbin(-99),
+            vp(-99),
+            format_errors(0),
+            dataword(-99){};
+
+      virtual ~GEM() = default;
+
+      inline void set_pad(const int bits) { pad = bits; }
```
Comment (rekovic) [inline DataFormats/L1TMuon/interface/EMTF/GEM.h:30]: please use the naming convention http://cms-sw.github.io/cms_coding_rules.html#2----naming-rules-1 for member variables, getters, and setters in this class.

File: DataFormats/L1TMuon/interface/EMTFHit.h [hunk 2]
```diff
@@ -80,16 +87,17 @@ namespace l1t {
...
 
     // void ImportCSCCorrelatedLCTDigi (const CSCCorrelatedLCTDigi& _digi);
     CSCCorrelatedLCTDigi CreateCSCCorrelatedLCTDigi() const;
     // void ImportRPCDigi (const RPCDigi& _digi);
     // RPCDigi CreateRPCDigi() const;
     // void ImportCPPFDigi (const CPPFDigi& _digi);
     CPPFDigi CreateCPPFDigi() const;
-    // void ImportGEMPadDigi (const GEMPadDigi& _digi);
-    // GEMPadDigi CreateGEMPadDigi() const;
+    // void ImportGEMPadDigiCluster (const GEMPadDigiCluster& _digi);  // TODO: implement placeholder when others are implemented
```
Comment (rekovic) [inline DataFormats/L1TMuon/interface/EMTFHit.h:99]: please remove commented-out code

File: DataFormats/L1TMuon/interface/EMTFHit.h [hunk 3]
```diff
@@ -80,16 +87,17 @@ namespace l1t {
...
+    GEMPadDigiCluster CreateGEMPadDigiCluster() const;
 
     // void PrintSimulatorHeader() const;
```
Comment (rekovic) [inline DataFormats/L1TMuon/interface/EMTFHit.h:102]: please remove commented-out code.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockGEM.cc [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+        if (!exact_duplicate)
+          res_GEM->insertDigi(Hit_.GEM_DetId(), Hit_.CreateGEMPadDigiCluster());
+
+        // Finished with unpacking one GEM Data Record
+        return true;
+
+      }  // End bool GEMBlockUnpacker::unpack
+
+      // bool GEMBlockPacker::pack(const Block& block, UnpackerCollections *coll) {
```
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockGEM.cc:247]: please remove commented-out code.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFCollections.cc [hunk 5]
```diff
@@ -30,13 +30,22 @@ namespace l1t {
           EMTFCPPFs_ZS_->push_back(c);
         }
       }
+      // TODO: Do we need something equivalent for GEMs - JS 03.07.20
+      // for (const auto& g : *EMTFGEMPadClusters_) {
```
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFCollections.cc:34]: Please remove this commented-out code.  It looks like it is a copy of some other code from elsewhere?  Is there going to be another PR which will implement an analogous code here ?

File: DataFormats/L1TMuon/interface/EMTF/GEM.h (hunk 1 above)
Comment (jsturdy) [reply DataFormats/L1TMuon/interface/EMTF/GEM.h:30]: This class was added based on the template of other classes in the `interface/EMTF` under the assumption that local continuity had higher importance than global style, not as an endorsement of any particular style choice.
Such an uniformization throughout the remainder of the code was left as a potential followup PR

File: DataFormats/L1TMuon/interface/EMTFHit.h (hunk 2 above)
Comment (jsturdy) [reply DataFormats/L1TMuon/interface/EMTFHit.h:99]: Additions were made to reflect structure of current code, and placeholder comments that were already present were added for the new portions

File: DataFormats/L1TMuon/interface/EMTFHit.h (hunk 3 above)
Comment (jsturdy) [reply DataFormats/L1TMuon/interface/EMTFHit.h:102]: Mentioned line was unmodified

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockGEM.cc (hunk 4 above)
Comment (jsturdy) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockGEM.cc:247]: Commented code reflects state of other EMTF block unpackers, and represents placeholders for the needed and future packer code

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFCollections.cc (hunk 5 above)
Comment (jsturdy) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFCollections.cc:34]: Indeed, if one expands the diff, the previous section has exactly this block for CPPF, and this comment is specifically to ensure that it's not forgotten by the developer to check whether such similar logic is *also* needed for the GEM digis.

**Comment by jsturdy** [comment]
I've left several replies, but as mentioned previously, I'm no longer on CMS, so I'll defer to @jiafulow to decide how to follow up on the requested cosmetic changes.

**Comment by boudoul** [comment]
Dear @rekovic and @jiafulow , could you please comment on the status of this PR ?

**Comment by rekovic** [comment]
comments are not yet addressed.

**Comment by jiafulow** [comment]
@rekovic That is simply not true. **All** the comments have been correctly addressed by Jared.

**Comment by rekovic** [comment]
If really needed to keep the comment.

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
please test
to refresh the tests

**Comment by silviodonato** [comment]
+1
